src/utils/IdealGraphVisualizer/HierarchicalLayout/src/main/java/com/sun/hotspot/igv/hierarchicallayout/HierarchicalStableLayoutManager.java

```diff
@@ -0,0 +1,1913 @@
+            }
+
+            if (pos < minX) {
+                n.x = minX;
+            } else {
+
+                LayoutNode rightNeighbor;
+                SortedSet<LayoutNode> tailSet = treeSet.tailSet(n);
+                int maxX = Integer.MAX_VALUE;
+                if (!tailSet.isEmpty()) {
+                    rightNeighbor = tailSet.first();
+                    maxX = rightNeighbor.x - offset(n, rightNeighbor) - n.width;
+                }
+
+                n.x = Math.min(pos, maxX);
+
+                assert minX <= maxX : minX + " vs " + maxX;
+            }
+
+            treeSet.add(n);
+        }
+    }
+
+    private class AssignYCoordinates {
+        void run() {
+
+            // Reset all values before assigning y-coordinates
+            for (LayoutNode n : nodes) {
+                if (n.vertex != null) {
+                    updateNodeWithReversedEdges(n);
+                } else {
+                    n.height = DUMMY_HEIGHT;
+                }
+                n.y = 0;
+            }
+
+            int curY = 0;
+
+            for (int i = 0; i < layers.size(); i++) {
+                List<LayoutNode> layer = layers.get(i);
+                int maxHeight = 0;
+                int baseLine = 0;
+                int bottomBaseLine = 0;
+                for (LayoutNode n : layer) {
+                    maxHeight = Math.max(maxHeight, n.height - n.yOffset - n.bottomYOffset);
+                    baseLine = Math.max(baseLine, n.yOffset);
+                    bottomBaseLine = Math.max(bottomBaseLine, n.bottomYOffset);
+                }
+
+                int maxXOffset = 0;
+                for (LayoutNode n : layer) {
+                    if (n.vertex == null) {
+                        // Dummy node
+                        n.y = curY;
+                        n.height = maxHeight + baseLine + bottomBaseLine;
+
+                    } else {
+                        n.y = curY + baseLine + (maxHeight - (n.height - n.yOffset - n.bottomYOffset)) / 2 - n.yOffset;
+                    }
+
+                    for (LayoutEdge e : n.succs) {
+                        int curXOffset = Math.abs(n.x - e.to.x);
+                        maxXOffset = Math.max(curXOffset, maxXOffset);
+                    }
+                }
+
+                curY += maxHeight + baseLine + bottomBaseLine;
+                curY += LAYER_OFFSET + ((int) (Math.sqrt(maxXOffset) * 1.5));
+            }
+        }
+    }
+
+    private class ComputeLayoutScore {
+        /**
+         * https://www.geeksforgeeks.org/check-if-two-given-line-segments-intersect/
+         *
+         * @param p
+         * @param q
+         * @param r
+         * @return true if point q lies on line segment 'pr'
+         */
+        private boolean onSegment(Point p, Point q, Point r) {
+            if (q.x <= Math.max(p.x, r.x) && q.x >= Math.min(p.x, r.x) &&
+                    q.y <= Math.max(p.y, r.y) && q.y >= Math.min(p.y, r.y))
```

Comment:
you can directly `return (q.x <= Math.max(p.x, r.x) && q.x >= Math.min(p.x, r.x) &&
                    q.y <= Math.max(p.y, r.y) && q.y >= Math.min(p.y, r.y))`